src/ecdh.rs

```diff
@@ -105,6 +122,44 @@ impl SharedSecret {
         ss.set_len(32); // The default hash function is SHA256, which is 32 bytes long.
         ss
     }
+
+    /// Creates a new shared secret from a pubkey and secret key with applied custom hash function
+    /// # Examples
+    /// ```
+    /// # use secp256k1::ecdh::SharedSecret;
+    /// # use secp256k1::{Secp256k1, PublicKey, SecretKey};
+    /// # fn sha2(_a: &[u8], _b: &[u8]) -> [u8; 32] {[0u8; 32]}
+    /// # let secp = Secp256k1::signing_only();
+    /// # let secret_key = SecretKey::from_slice(&[3u8; 32]).unwrap();
+    /// # let secret_key2 = SecretKey::from_slice(&[7u8; 32]).unwrap();
+    /// # let public_key = PublicKey::from_secret_key(&secp, &secret_key2);
+    ///
+    /// let secret = SharedSecret::new_with_hash(&public_key, &secret_key, |x,y| {
+    ///     let hash: [u8; 32] = sha2(&x,&y);
+    ///     hash.into()
+    /// });
+    ///
+    /// ```
+    pub fn new_with_hash<F>(point: &PublicKey, scalar: &SecretKey, mut hash_function: F) -> SharedSecret
+        where F: FnMut([u8; 32], [u8; 32]) -> SharedSecret
+    {
+        let mut ss = SharedSecret::empty();
+        let hashfp: ffi::EcdhHashFn = hash_callback::<F>;
```

Comment:
This is really a fascinating use of Rust's type system. I did not know that generics and FFI would be allowed to interact like this.